### ISSUE TITLE
0.1 to 0.4 upgrade big slowdown

### ISSUE THREAD

**Issue by Redsandro** [Leonidas-from-XIV/node-xml2js#157]
Hi people,

I have been using `node-xml2js` v`0.1.14`'s `xml2json.Parser().parseString` without further options on `XML` files which kept getting bigger. Now these files up to ~2MB take nearly a second to translate to `js`.

So I thought, let's update and see if there are speed improvements. However, after I updated `node-xml2js` to v`0.4.4`, I noticed that in stead of a speed increase, I get a big slowdown were it now takes 8 to 10 seconds.

Are there some special options necessary in the 0.4 version? Or is this module just not meant for bigger XML? Something else I'm missing?

For now I'm downgrading to the much faster version 0.1, and I'll continue my search for fast converters.

**Comment by csimi** [comment]
If anyone is interested, easysax seems to be a damn fast parser, written in JS.
I had to fix some bugs in easysax myself and integrate it into xml2js but it sped up things pretty well.

```
sax x 57,812 ops/sec ±7.41% (78 runs sampled)
node-xml x 76,807 ops/sec ±1.75% (87 runs sampled)
libxmljs x 163,375 ops/sec ±2.58% (88 runs sampled)
node-expat x 201,663 ops/sec ±0.76% (84 runs sampled)
easysax x 828,169 ops/sec ±2.59% (86 runs sampled)
```

I feel there has to be a catch somewhere but I don't see it yet.

**Comment by Redsandro** [comment]
Seems to be streaming. Doesn't convert to object itself. 
(Read: Not inline-replacable with node-xml2js)
Am I wrong?

**Comment by csimi** [comment]
Yeah, this is regarding to the discussion about the sax-js "backend" of xml2js and replacing it with node-expat.
I've seen a drop in CPU usage of around 50% after replacing sax-js with easysax. Much of the time is still spent on building an object of the whole XML so I'm thinking about just straight-up using the SAX parser.
We're saturating our 100Mbps proxy vms with gzipped XML files so for parsing that much of data every little speedup counts.

**Comment by tflanagan** [comment]
How does easysax work with browserify?

**Comment by Leonidas-from-XIV** [comment]
@csimi Does your version with the easysax backend pass the unit tests? I am in no means married to sax-js, I just want to avoid a dependency that has to be compiled.

On the other hand, if you saturate your network connection, serializing XML into an object is probably going to be inherently expensive, if you have high performance in mind a streaming solution like raw SAX or similar might indeed be preferable.

**Comment by tflanagan** [comment]
@Leonidas-from-XIV, if you replace the parser with one that does not work 100% with browserify, then I will be forced to fork it.

Can we introduce an external hook rather than outright replacing it? Side effects could be huge - If one of these replacement libs is actually async, unlike sax-js (because of eventemitter), then a lot of people will come screaming.

**Comment by Leonidas-from-XIV** [comment]
@tflanagan I understand your issue and will try to take it into account.

Your complaint is actually why I would be agains supporting multiple backends, because then the semantics of the library might change in unforeseen ways and some people will be surprised trying to have the same behaviour everywhere might end up an uphill battle for little benefit. I'd rather have one solid backend that works for all.

**Comment by kyrylkov** [comment]
@csimi can you publish your xml2js + easysax repo?

**Comment by kyrylkov** [comment]
My run on a 46k XML file:
```
sax x 66.98 ops/sec ±2.62% (50 runs sampled)
node-xml x 70.33 ops/sec ±2.56% (54 runs sampled)
libxmljs x 162 ops/sec ±2.46% (62 runs sampled)
node-expat x 116 ops/sec ±3.50% (59 runs sampled)
ltx x 236 ops/sec ±3.67% (60 runs sampled)
EasySax x 1,167 ops/sec ±4.62% (59 runs sampled)
Fastest is EasySax
```

**Comment by Redsandro** [comment]
@kyrylkov just for comparison, can you add [RapidX2J](https://www.npmjs.com/package/rapidx2j) to this comparison? It will illustrate using RapidXML backend.

**Comment by kyrylkov** [comment]
@Redsandro It doesn't seem to compile with Node.js 5.8.0?  Does it support Node.js 4.x and 5.x?

**Comment by Redsandro** [comment]
@kyrylkov oops no idea actually. I'm running legacy (pre io.js-post-fork-merge) node.js for production reasons.

**Comment by Redsandro** [comment]
@kyrylkov Compiles on __4.x__ according to dev:

https://github.com/damirn/rapidx2j/issues/17

> it should work with node 4.4.0:
> 
> ```
git clone https://github.com/damirn/rapidx2j.git
npm install rapidx2j/
```
>
> I just tried it on mac os x w/o issues

**Comment by csimi** [comment]
I'll try to put something usable together during the weekend.
More benchmarks:

xml2js x 2.33 ops/sec ±27.49% (10 runs sampled)
xml2js easysax x 11.31 ops/sec ±19.67% (18 runs sampled)
rapidx2j x 28.87 ops/sec ±16.22% (30 runs sampled)
easysax x 129 ops/sec ±15.94% (30 runs sampled)

The easysax bench doesn't actually build a js object, just runs a SAX pass over my 215KiB XML test file (fastest mode, without parsing attributes, etc).

Rapidx2j is pretty fast but just how much time it takes to actually build a JS object is clearly visible.

I feel like EasySax is too fast (compared to expat for example) to be standards compliant. There has to be a catch somewhere.

**Comment by Redsandro** [comment]
> how much time it takes to actually build a JS object is clearly visible.

How? I take any parser and it just takes time. 

Is there a (simple) way to use EasySax for getting the JS object out of xml? I'd like to see if it can work for me. But in the docs it seems to just throw events on every element encountered (that's what I assumed was streaming, as SAX works like that too).

So far, rapidx2j seems the fastest but I cannot use it without forking, as the current module implements a custom lossy standard, changing caps and such.
